exporters/otlp/include/opentelemetry/exporters/otlp/otlp_http_exporter.h

```diff
@@ -0,0 +1,113 @@
+constexpr char kDefaultTracePath[] = "/v1/traces";
+// The default URL path to post metric data.
+constexpr char kDefaultMetricsPath[] = "/v1/metrics";
+// The default URL path to post metric data.
+constexpr char kDefaultLogPath[] = "/v1/logs";
+// The HTTP header "Content-Type"
+constexpr char kHttpContentType[] = "application/json";
+
+enum class BytesMappingKind
+{
+  kHexId,
+  kBase64,
+  kHex,
+};
+
+/**
+ * Struct to hold OTLP exporter options.
+ */
+struct OtlpHttpExporterOptions
+{
+  // The endpoint to export to. By default
+  // @see
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md
+  // @see https://github.com/open-telemetry/opentelemetry-collector/tree/main/receiver/otlpreceiver
+  std::string url = std::string("http://localhost:4317") + kDefaultTracePath;
+
+  // TODO: By default when false, set CURLOPT_SSL_VERIFYPEER to false
+  // bool use_ssl_credentials = false;
+  // If convert bytes into hex. By default, we will convert bytes into base64
+  BytesMappingKind json_bytes_mapping = BytesMappingKind::kHexId;
+
+  // If using the json name of protobuf field to set the key of json. By default, we will use the
+  // field name just like proto files.
+  bool use_json_name = false;
+
+  // Whether to print the status of the exporter in the console
+  bool console_debug = false;
+
+  // Maximum time to wait for response after sending http request(milliseconds)
+  int response_timeout = 30000;
+};
+
+/**
+ * The OTLP exporter exports span data in OpenTelemetry Protocol (OTLP) format.
+ */
+class OtlpHttpExporter final : public opentelemetry::sdk::trace::SpanExporter
+{
+public:
+  /**
+   * Create an OtlpHttpExporter using all default options.
+   */
+  OtlpHttpExporter();
+
+  /**
+   * Create an OtlpHttpExporter using the given options.
+   */
+  OtlpHttpExporter(const OtlpHttpExporterOptions &options);
+
+  /**
+   * Create a span recordable.
+   * @return a newly initialized Recordable object
+   */
+  std::unique_ptr<sdk::trace::Recordable> MakeRecordable() noexcept override;
+
+  /**
+   * Export
+   * @param spans a span of unique pointers to span recordables
+   */
+  sdk::common::ExportResult Export(
+      const nostd::span<std::unique_ptr<sdk::trace::Recordable>> &spans) noexcept override;
+
+  /**
+   * Shut down the exporter.
+   * @param timeout an optional timeout, the default timeout of 0 means that no
+   * timeout is applied.
+   * @return return the status of this operation
+   */
+  bool Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+private:
+  // Stores if this exporter had its Shutdown() method called
+  bool is_shutdown_ = false;
+
+  // The configuration options associated with this exporter.
+  const OtlpHttpExporterOptions options_;
+
+  // Object that stores the HTTP sessions that have been created
+  std::unique_ptr<opentelemetry::ext::http::client::curl::HttpOperation> curl_operation_;
```

Comment:
Any specific reason for directly using curl implementation, and not the http_client as we use in zipkin: https://github.com/open-telemetry/opentelemetry-cpp/blob/f793e6e45e181e8224e38871b837746b878bf821/exporters/zipkin/src/zipkin_exporter.cc#L6

This will allow using platform-specific HTTP client implementation - e.g, WinSock on Windows.

Reply:
I want to create session/request from URL directly, without split it into host and port and support HTTPS. The API of `Session` and `HttpClient` have not such APIs by now  and I don't want to use `HttpClientSync`.
I think we can switch to `HttpClient` when it supports this feature.

Reply:
Really sorry if I am missing something here, as I still have some doubts:

> I don't want to use HttpClientSync

What is the benefit of an async request here? As `SpanExporter::Export()` doesn't support callback, so we need to anyway wait to get a response before returning from this method. Also, `libcurl` doesn't have async api.

> support HTTPS 

I think both `HttpClient` and `HttpClientSync` support HTTPS, as the authentication data can be added as part of HTTP headers. Also, with `HttpClientSync` we don't need to split host and port.

Unless really required, it's better not to force application owners to install the `libcurl` library. Applications running on Windows/macOS may prefer using native libraries for network communication.  Plan is to provide [HttpClientFactory](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/ext/include/opentelemetry/ext/http/client/http_client_factory.h) implementation for these native libraries too similar to [curl](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/ext/src/http/client/curl/http_client_factory_curl.cc). And let the HttpClient library pick up implementation as available for that platform.

Reply:
> What is the benefit of an async request here? As SpanExporter::Export() doesn't support callback, so we need to anyway wait to get a response before returning from this method. Also, libcurl doesn't have async api.

Just like `ElasticsearchLogExporter` , using async request allow us to cancle the request and exit gracefully here. libcurl has [multi interface](https://curl.se/libcurl/c/libcurl-multi.html) that can be used to perform asynchronous requests. I have some project use libuv and libcurl to send HTTP requests asynchronously.(e.g. [http_request](https://github.com/atframework/atframe_utils/blob/main/src/network/http_request.cpp) )

> support HTTPS

The codes in `opentelemetry::ext::http::client::curl::Session` is

```cpp
public:
  Session(HttpClient &http_client, const std::string &host, uint16_t port = 80)
      : http_client_(http_client), is_session_active_(false)
  {
    if (host.rfind("http://", 0) != 0 && host.rfind("https://", 0) != 0)
    {
      host_ = "http://" + host;  // TODO - https support
    }
    else
    {
      host_ = host;
    }
    host_ += ":" + std::to_string(port) + "/";
  }
```

It seems to reset `https://` to `http://`. 

I also think it's better to switch to `HttpClient` after HttpClient has API of create request only by URL. Or shall I use `HttpClientSync` for now and switch to `HttpClient` later?

Reply:
> Just like ElasticsearchLogExporter , using async request allow us to cancle the request and exit gracefully here

`ElasticsearchLogExporter` was implemented when sync request support was missing in the HttpClient library ( https://github.com/open-telemetry/opentelemetry-cpp/pull/444#discussion_r542112168 ), and there was a plan to migrate it on sync request. But I agree request cancellation is something we can use with the async request.


> I also think it's better to switch to HttpClient after HttpClient has API of create request only by URL. Or shall I use HttpClientSync for now and switch to HttpClient later?

I would suggest using `HttpClientSync` for now instead of `libcurl` directly. I agree `HttpClient` does need changes for supporting HTTPS and full-URL, and I will add a ticket for that. We can move to that once it's fixed.

Reply:
> I would suggest using HttpClientSync for now instead of libcurl directly. I agree HttpClient does need changes for supporting HTTPS and full-URL, and we can create a ticket for it. And then move to that once fixed.

Done. And rename `OtlpExporter`, `OtlpExporterOptions`, `otlp_exporter.h`, `otlp_exporter.cpp` to `OtlpGrpcExporter`, `OtlpGrpcExporterOptions`, `otlp_grpc_exporter.h`, `otlp_grpc_exporter.cpp` .